scanpy/tools/_tsne.py

```diff
@@ -75,48 +94,77 @@ def tsne(
     **X_tsne** : `np.ndarray` (`adata.obs`, dtype `float`)
         tSNE coordinates of data.
     """
-    start = logg.info('computing tSNE')
     adata = adata.copy() if copy else adata
-    X = _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)
-    # params for sklearn
-    params_sklearn = dict(
-        perplexity=perplexity,
-        random_state=random_state,
-        verbose=settings.verbosity > 3,
-        early_exaggeration=early_exaggeration,
+
+    if neighbors_key is None:
+        neighbors_key = "neighbors"
+
+    if neighbors_key not in adata.uns:
+        raise ValueError(
+            f"Did not find .uns['{neighbors_key}']. Run `sc.pp.neighbors` first."
+        )
+
+    start = logg.info("computing tSNE")
+
+    tsne_params = dict(
         learning_rate=learning_rate,
+        early_exaggeration_iter=early_exaggeration_iter,
+        early_exaggeration=early_exaggeration,
+        n_iter=n_iter,
+        exaggeration=exaggeration,
+        dof=dof,
+        initial_momentum=initial_momentum,
+        final_momentum=final_momentum,
+        max_step_norm=max_step_norm,
+        negative_gradient_method=negative_gradient_method,
     )
+    adata.uns["tsne"] = {"params": dict(tsne_params)}
+
+    verbose = settings.verbosity > 3
+
+    neighbors = NeighborsView(adata, neighbors_key)
+    connectivities = neighbors["connectivities"]
+
+    # If `binarize="auto"`, we binarize everything except t-SNE weights
+    if binarize == "auto":
+        binarize = neighbors["params"]["method"] != "tsne"
```

Comment:
Binarization will produce uniform weights that differ ever so slightly differently from `openTSNE.affinity.Uniform`. The difference is that in openTSNE, we do not have a symmetric graph, so when we symmetrize uniform weights, the points that appear as each others neighbors actually have higher weights. So in that case we have two sets of possible weights, e.g. 0.05 and 0.1.

Here, we already have a symmetric connectivity graph, so this doesn't happen, and all points are assigned equal weights. I strongly suspect this has practically no impact on the resulting embeddings.

Reply:
I agree. For consistency, we can make the `Uniform` affinity class in openTSNE to support not only `symmetrize=True/False` but also something like `symmetrize='or'` which would compute `P=(P+P.T) > 0`. I thought about it a while ago already actually.